### PR TITLE
Add Skip and TrySkip APIs to Utf8JsonReader with tests.

### PR DESCRIPTION
**Pull request by ahsonkhan**
Fixes https://github.com/dotnet/corefx/issues/33295

cc @bartonjs, @steveharter, @JeremyKuhne, @maryamariyan

### CODE REVIEW

**Comment by ahsonkhan** [comment]
The outerloop test failures are unrelated to this PR. I accidentally triggered them.

File: src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.cs [hunk 1]
```diff
@@ -260,6 +260,85 @@ public bool Read()
...
 
+        /// <summary>
+        /// Skips the children of the current JSON token.
+        /// </summary>
+        /// <exception cref="InvalidOperationException">
+        /// Thrown when the reader was given partial data with more data to follow (i.e. _isFinalBlock is false).
+        /// </exception>
+        public void Skip()
+        {
+            if (!_isFinalBlock)
```
Comment (bartonjs) [inline src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.cs]: Is there any way that a caller who receives an already open Utf8JsonReader can know if it's in this state?  I don't see a property exposing this value...

File: src/System.Text.Json/tests/JsonTestHelper.cs [hunk 2]
```diff
@@ -260,6 +260,90 @@ public static string ListToString(List<object> list)
...
+                case JsonToken.Date:
+                case JsonToken.Bytes:
+                case JsonToken.Undefined:
+                case JsonToken.Raw:
+                default:
+                    throw new InvalidOperationException();
+            }
+        }
+
+        public static string InsertCommentsEverywhere(string jsonString)
```
Comment (bartonjs) [inline src/System.Text.Json/tests/JsonTestHelper.cs]: Maybe I'm missing something, but this test doesn't seem related to the code change

File: src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.cs (hunk 1 above)
Comment (ahsonkhan) [reply src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.cs]: > Is there any way that a caller who receives an already open Utf8JsonReader can know if it's in this state?

No, there isn't.

We can expose it, but do you think adding this API requires it (i.e. we can't have one without the other)?

File: src/System.Text.Json/tests/JsonTestHelper.cs (hunk 2 above)
Comment (ahsonkhan) [reply src/System.Text.Json/tests/JsonTestHelper.cs]: We use it in `TestTrySkipWithComments`.

**Comment by bartonjs** [comment]
Two tests that I think would be good that I don't think I saw.  But, I may have just missed them.

* Call Skip / TrySkip as the first thing after restoring state into a new reader.  Might be good to add one, just to show that the async usage pattern will succeed.  (The usage could have been isFinal: true on both sides of the await... it might be awaiting something as a result of the payload, not awaiting part of it)
* Call TrySkip on a partial where it ends in a whitespace run, e.g. "[ 1, |       2 ]" when the current value is the "1" and the pipe is where the non-final buffer is sliced.
  * Presumably that returns false, since it doesn't know where the next token starts?

File: src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.cs (hunk 1 above)
Comment (bartonjs) [reply src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.cs]: I think adding this API exposes a need for it.  Library authors wouldn't know that it's safe (or not) to use this method.  It's a precondition exception that they can't check.

**Comment by ahsonkhan** [comment]
> Two tests that I think would be good that I don't think I saw

I have tests similar to the cases you mentioned, but will add simple ones that test it more directly.

**Comment by ahsonkhan** [comment]
Unrelated `System.Drawing.Common.Tests` failure on RedHat:
https://mc.dot.net/#/user/dotnet-bot/pr~2Fdotnet~2Fcorefx~2Frefs~2Fpull~2F37793~2Fmerge/test~2Ffunctional~2Fcli~2Finnerloop~2F/20190527.20

File: src/System.Text.Json/src/Resources/Strings.resx [hunk 3]
```diff
@@ -362,6 +362,9 @@
   <data name="UnexpectedEndOfDataWhileReadingComment" xml:space="preserve">
     <value>Unexpected end of data while reading a comment.</value>
   </data>
+  <data name="CannotSkip" xml:space="preserve">
+    <value>Cannot skip tokens on partial JSON. Either get the whole payload and set _isFinalBlock to true or call TrySkip.</value>
```
Comment (bartonjs) [inline src/System.Text.Json/src/Resources/Strings.resx]: Why is there an underscore in the resource string?  The caller can't set the field.   You either mean the parameter (no _) or the property (no _, capital letter)

File: src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.cs [hunk 4]
```diff
@@ -260,6 +267,102 @@ public bool Read()
             return retVal;
         }
 
+        /// <summary>
+        /// Skips the children of the current JSON token.
+        /// </summary>
+        /// <exception cref="InvalidOperationException">
+        /// Thrown when the reader was given partial data with more data to follow (i.e. _isFinalBlock is false).
```
Comment (bartonjs) [inline src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.cs]: _isFinalBlock => IsFinalBlock

File: src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.cs [hunk 5]
```diff
@@ -260,6 +267,102 @@ public bool Read()
...
+        /// </exception>
+        public void Skip()
```
Comment (bartonjs) [inline src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.cs]: Consider adding something like
```C#
/// <remarks>
/// When <see cref="TokenType"/> is <see cref="JsonTokenType.PropertyName" /> the reader first moves to the property value.
/// When <see cref="TokenType"/> (originally, or after advancing) is <see cref="JsonTokenType.StartObject" /> or 
/// <see cref="JsonTokenType.StartArray" /> the reader advances to the matching
/// <see cref="JsonTokenType.EndObject" /> or <see cref="JsonTokenType.EndArray" />.
///
/// For all other token types the reader does not move. After the next call to <see cref="Read"/> the reader will be at
/// the next value (when in an array) or the next property name (when in an object).
/// </remarks>

File: src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.cs [hunk 6]
```diff
@@ -260,6 +267,102 @@ public bool Read()
             return retVal;
         }
 
+        /// <summary>
+        /// Skips the children of the current JSON token.
+        /// </summary>
+        /// <exception cref="InvalidOperationException">
```
Comment (bartonjs) [inline src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.cs]: It also throws JsonReaderException (if that's still the type name) if the JSON is invalid... so anything that Read says it would throw.

File: src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.cs [hunk 7]
```diff
@@ -260,6 +267,102 @@ public bool Read()
...
+        }
+
+        /// <summary>
+        /// Tries to skip the children of the current JSON token.
+        /// </summary>
+        /// <returns>True if there was enough data for the children to be skipped successfully, else false.</returns>
+        /// <remarks>
+        /// If the reader did not have enough data to completely skip the children of the current token,
+        /// it will be reset to the state it was in before the method was called.
+        /// </remarks>
```
Comment (bartonjs) [inline src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.cs]: Presumably it throws on invalid JSON, instead of resetting state.  So those exceptions are probably worth mentioning.

File: src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.cs (hunk 5 above)
Comment (ahsonkhan) [reply src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.cs]: Funnily, I had considered adding this :)
